x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java

```diff
@@ -119,39 +145,68 @@ public static Map<RealmConfig.RealmIdentifier, Settings> getRealmSettings(Settin
      * Computes the realm name to domain name association.
      * Also verifies that realms are assigned to at most one domain and that domains do not refer to undefined realms.
      */
-    public static Map<String, String> computeRealmNameToDomainNameAssociation(Settings globalSettings) {
-        final Set<RealmConfig.RealmIdentifier> allRealmIdentifiers = RealmSettings.getRealmSettings(globalSettings).keySet();
-        final Map<String, Set<String>> realmToDomainsMap = new HashMap<>();
-        for (String domainName : DOMAIN_TO_REALM_ASSOC_SETTING.getNamespaces(globalSettings)) {
+    public static Map<String, DomainConfig> computeRealmNameToDomainConfigAssociation(Settings globalSettings) {
+        final Settings domainSettings = globalSettings.getByPrefix(DOMAIN_SETTING_PREFIX);
+        final Map<String, Set<DomainConfig>> realmToDomainsMap = new HashMap<>();
+        for (String domainName : domainSettings.names()) {
             if (domainName.startsWith(RESERVED_REALM_AND_DOMAIN_NAME_PREFIX)) {
                 throw new IllegalArgumentException(
                     "Security domain name must not start with \"" + RESERVED_REALM_AND_DOMAIN_NAME_PREFIX + "\""
                 );
             }
-            Setting<List<String>> realmsByDomainSetting = DOMAIN_TO_REALM_ASSOC_SETTING.getConcreteSettingForNamespace(domainName);
-            for (String realmName : realmsByDomainSetting.get(globalSettings)) {
-                realmToDomainsMap.computeIfAbsent(realmName, k -> new TreeSet<>()).add(domainName);
+
+            final Setting<List<String>> domainRealmsSetting = DOMAIN_TO_REALM_ASSOC_SETTING.getConcreteSettingForNamespace(domainName);
+            if (false == domainRealmsSetting.exists(globalSettings)) {
+                throw new IllegalArgumentException("[" + domainRealmsSetting.getKey() + "] must exist for security domain configuration");
+            }
+            final List<String> memberRealmNames = domainRealmsSetting.get(globalSettings);
+            // TODO: Does it make sense to have empty realms for a domain?
```

Comment:
So far we allow an empty list of realms to be defined for a domain. But it seems too much leniency. Such domain is also not tracked in current code because this method builds a map of realm to domain. If the domain has no realm, it is not included in the map. It still works because other parts of the code should never need to know about the domain. But it still feels weird to some extent. I didn't change it in this PR. But we can discuss whether this should be fixed as a bug before 8.3 is out.